File: WINGS_AI_Bootcamp/Elixir/Pascal_Triangle.ex
defmodule Pascal do
  def pascals_triangle(n) do
    Stream.iterate([1], &next_row/1)
    |> Enum.take(n)
  end

  defp next_row(row) do
    left = [0 | row]
    right = row ++ [0]

    Enum.zip(left, right)
    |> Enum.map(fn {a, b} -> a + b end)
  end

  def print_pascals_triangle(n) do
    triangle = pascals_triangle(n)
    last_row_str = Enum.join(List.last(triangle), "   ")
    max_width = String.length(last_row_str)

    Enum.each(triangle, fn row ->
      row_str = Enum.join(row, "   ")
      padding = String.duplicate(" ", div(max_width - String.length(row_str), 2))
      IO.puts(padding <> row_str)
    end)
  end
end

Pascal.print_pascals_triangle(6)

File: WINGS_AI_Bootcamp/Python/Pascal_Triangle.py
# Generate Pascals Triangle
'''
                            1
                        1       1
                    1       2       1
                1       3       3       1
            1       4       6       4       1
        1       5       10      10      5       1

'''
def pascals_triangle(n: int):
    triangle = []
    for i in range(n):
        row = [1]
        if triangle:
            row += [triangle[-1][j] + triangle[-1][j+1] for j in range(len(triangle[-1]) - 1)] + [1]
        triangle.append(row)
    return triangle

def print_pascals_triangle(n: int):
    triangle = pascals_triangle(n)
    max_width = len('   '.join(map(str, triangle[-1])))  # width of last row string

    for row in triangle:
        row_str = '   '.join(map(str, row))
        print(row_str.center(max_width))

print_pascals_triangle(5)